### PR TITLE
Check that abstract methods are implemented. Fixes #835

### LINKED ISSUE #835: Unimplemented abstract methods don't raise error unless called
```crystal
abstract class A
  abstract def a
end

class B < A
end
```
Compiles just fine even though B doesn't implement A's abstract method a. The error only shows up when you call a on B's instance. In my opinion, when a class inherits from another class, it should be obliged to implement it's methods. The current behavior may lead to bugs going undetected especially for libraries (where methods may not be called on every object in the library itself).
I know this may go against the language's philosophy and would like to hear opinions.

### PR DESCRIPTION
**Pull request by asterite**
This makes the compiler check that abstract methods are implemented in subclasses (or including classes in the case of modules).

This change uncovered a couple of "bugs":
* `Object#hash` is abstract but `StaticArray` and `Proc` didn't implement it
* `Int#popcount` is abstract but `BigInt` didn't implement it

The more the compiler can help us catch these errors, the better.

With this, errors such as [this one](https://github.com/manastech/crystal/issues/1975#issuecomment-167577053) should be easier to catch: we could make `IO#close` abstract and require subclasses to implement it. Same goes with `IO#flush`, `IO#sync?`, etc. (although another alternative is to define those methods as no-op).

This PR has some TODOs: abstract methods for generic types aren't checked, and if the methods have splats or default arguments they aren't checked either. We'll add these checks in the future (they are a bit more tricky but doable).

### CODE REVIEW

File: src/proc.cr [hunk 1]
```diff
@@ -35,6 +35,10 @@ struct Proc
     call(other)
   end
 
+  def hash
+    object_id.hash
+  end
```
Comment (jhass) [inline src/proc.cr]: Shouldn't `Object` provide this as a default implementation?

Comment (asterite) [inline src/proc.cr]: Reference does, but structs don't have an object_id

Comment (jhass) [inline src/proc.cr]: Oh I see, thanks

Comment (asterite) [inline src/proc.cr]: We could make Value's hash be computed from the raw bytes, but maybe it's better to ensure that each subclass defines a correct logic.

**Pull request merged**